### PR TITLE
fix alias bug + refactor

### LINKED ISSUE #511: Task unmarshaling fails with aliases
**Describe the bug**

**To Reproduce**

Run this task:

```
version: v1.1.0
alias:
  values:
    repo: myrepo
steps:
  - build: -t $Registry/$repo:$ID  -t $Registry/hello-world:$ID -f Dockerfile .
  - push:
    - "$Registry/hello-world:$ID"
```

1. 
2. 
3. 

**Expected behavior**

**Screenshots**

**Any relevant environment information**
 - OS: [e.g. Windows]
 - Version [e.g. v1.0.0]

**Additional context**

### PR DESCRIPTION
**Pull request by shahzzzam**
**Purpose of the PR**

Re-ordered steps for exec.

Instead of: 
1. Preprocess task
2. Unmarshal task
3. Expand command aliases
4. Marshal task
5. Render gotemplates
6. Unmarshal Task for run

we simplify to this:
1. Preprocess task
2. Render gotemplates
3. Unmarshal task for run
4. Expand command aliases

Also as a part of #501 

Fixes #511

### CODE REVIEW

File: cmd/acb/commands/exec/exec.go [hunk 1]
```diff
@@ -201,55 +199,49 @@ var Command = cli.Command{
...
+		// Find a way to do "rendering" and "preprocessing" in Task initialization `graph` module, instead of `commands`.
 		if shouldIncludeAlias {
 			log.Printf("Alias support enabled for version >= 1.1.0, please see https://aka.ms/acr/tasks/task-aliases for more information.")
-			// Generate the base task file without resolving environment variables.
-			task, err = graph.NewTaskFromBytes(template.GetData(), true)
-			if err != nil {
-				return err
+			// Preprocess the task to replace all aliases based on the alias sources.
+			post, aliasPtr, _, aliasErr := graph.PreprocessBytes(template.GetData())
+			alias = aliasPtr
```
Comment (shahzzzam) [inline cmd/acb/commands/exec/exec.go]: @djyou @juliusl @mnltejaswini  golang experts, Question about scope :
using 
```
post, alias, _, aliasErr := graph.PreprocessBytes(template.GetData())
```

would give me a compile error `alias declared and not used`

which is why this extra assignment. LMK if theres some idiomatic way out there :)

**Comment by northtyphoon** [comment]
If you put Expand command aliases at the last step and the alias value references a go template, how does the go template get rendered?

File: cmd/acb/commands/exec/exec.go (hunk 1 above)
Comment (northtyphoon) [reply cmd/acb/commands/exec/exec.go]: For the particular issue, you can declare other variables outside if block along with alias. So you don't need to use := and avoid variable shadowing.

---
In reply to: [326838074](https://github.com/Azure/acr-builder/pull/510#discussion_r326838074) [](ancestors = 326838074)

File: go.mod [hunk 2]
```diff
@@ -32,3 +32,5 @@ require (
 	gopkg.in/yaml.v2 v2.2.2
 	gotest.tools v2.2.0+incompatible // indirect
 )
+
+go 1.13
```
Comment (northtyphoon) [inline go.mod]: Is it auto added by the tool? If 1.13 is required, make sure you update Dockerfile to 1.13 as well.

File: cmd/acb/commands/exec/exec.go [hunk 3]
```diff
@@ -201,55 +199,49 @@ var Command = cli.Command{
...
 		}
 
+		// Add all creds provided by the user in the --credential flag
+		credentials, err := graph.CreateRegistryCredentialFromList(creds)
+		if err != nil {
+			return err
+		}
+
 		versionInUse := graph.FindVersion(template.GetData())
 		shouldIncludeAlias := versionInUse == "" || versionInUse >= "v1.1.0"
```
Comment (northtyphoon) [inline cmd/acb/commands/exec/exec.go]: > versionInUse == "" [](start = 23, length = 19)

As we discussed, let's only opt-in v1.1.0 version for now.

**Comment by shahzzzam** [comment]
abandoning now

**Comment by shahzzzam** [comment]
do not delete the branch

**Pull request closed without merging**